### ISSUE TITLE
ParquetConvert.DeserializeAsync throw Exception System.ArgumentOutOfRangeException  for null DateTime

### ISSUE THREAD

**Issue by vladyslav-panasenko** [aloneguid/parquet-dotnet#224]
**Library Version:** latest: 4.2.3
**.NET Version:** .NET 6
**OS:** Windows 10

#### Expected Behaviour
ParquetConvert.DeserializeAsync will not throw an Exception System.ArgumentOutOfRangeException for null DateTime under specific conditions

#### Actual Behaviour
ParquetConvert.DeserializeAsync throws Exception System.ArgumentOutOfRangeException for null DateTime under specific conditions

#### Exception
```json

{
  "ClassName": "System.ArgumentOutOfRangeException",
  "Message": "Year, Month, and Day parameters describe an un-representable DateTime.",
  "Data": null,
  "InnerException": null,
  "HelpURL": null,
  "StackTraceString": "   at System.DateTimeOffset..ctor(Int32 year, Int32 month, Int32 day, Int32 hour, Int32 minute, Int32 second, TimeSpan offset)\r\n   at Parquet.File.Values.Primitives.NanoTime.op_Implicit(NanoTime nanoTime)\r\n   at Parquet.Data.ParquetEncoder.Decode(Stream source, Span`1 data, SchemaElement tse)\r\n   at Parquet.Data.ParquetEncoder.Decode(Array data, Int32 offset, Int32 count, SchemaElement tse, Stream source, Int32& elementsRead)\r\n   at Parquet.File.DataColumnReader.ReadColumn(BinaryReader reader, Encoding encoding, Int64 totalValues, Int32 maxReadCount, ColumnRawData cd)\r\n   at Parquet.File.DataColumnReader.ReadDataPage(PageHeader ph, ColumnRawData cd, Int64 maxValues)\r\n   at Parquet.File.DataColumnReader.ReadAsync(CancellationToken cancellationToken)\r\n   at Parquet.Extensions.TaskExtensions.SequentialWhenAll[T](IEnumerable`1 collection)\r\n   at Parquet.ParquetConvert.ReadAndDeserializeByRowGroupAsync[T](Int32 rowGroupIndex, ParquetReader reader, DataField[] dataFields)\r\n   at Parquet.ParquetConvert.DeserializeAsync[T](Stream input, Int32 rowGroupIndex, ParquetSchema fileSchema)\r\n   at Program.<Main>$(String[] args) in C:\\Harper\\parquet te\\ConsoleApp1\\ConsoleApp1\\Program.cs:line 24",
  "RemoteStackTraceString": null,
  "RemoteStackIndex": 0,
  "ExceptionMethod": null,
  "HResult": -2146233086,
  "Source": "System.Private.CoreLib",
  "WatsonBuckets": null,
  "ParamName": null,
  "ActualValue": null
}

```

#### Details
The code snippet that throws an exception is below.

With 8000 items that have DateTime = null deserialization works excellently. No exceptions.
With 7000 I got an exception.

Code failing at NanoTime.cs at line 78. 
https://github.com/aloneguid/parquet-dotnet/blob/781656222a7008f870c7bdffc22b82254dd972e3/src/Parquet/File/Values/Primitives/NanoTime.cs#L78

Code tries to create a new DateTimeOffset instance using negative day and month. 
![negative_date](https://user-images.githubusercontent.com/25965770/212837501-184bcbdf-2b71-4aa1-abb4-753e3f739886.jpg)

At line 1067 in class ParquetEncoder ArrayPool<byte>.Shared.Rent returns an array with non-zero values. 

```csharp
                    byte[] buf = ArrayPool<byte>.Shared.Rent(NanoTime.BinarySize * data.Length);
                    try {
                        Span<byte> span = buf.AsSpan(0, NanoTime.BinarySize * data.Length);
                        Read(source, span);
                        int offset = 0;
                        int els = 0;
                        while(offset + NanoTime.BinarySize <= span.Length) {
                            data[els++] = new NanoTime(span.Slice(offset));
                            offset += NanoTime.BinarySize;
                        }
                        return els;
                    }
```
https://github.com/aloneguid/parquet-dotnet/blob/bf24f0de55418190925e2febde8f57ad2bbcb36c/src/Parquet/Data/ParquetEncoder.cs#L1067
...

#### Code Snippet to reproduce 

```csharp

// test steps.
var data = Enumerable.Range(0, 10000).Select(i => new Data { Id = i, DateTime = DateTime.Now }).ToList();

foreach (var item in data.Take(7000))
{
    item.DateTime = null;
}

var stream = new MemoryStream();

var schema = await ParquetConvert.SerializeAsync(data, stream);

try
{
    var result = await ParquetConvert.DeserializeAsync<Data>(stream);
}
catch (Exception e)
{
    var res = JsonConvert.SerializeObject(e);
    Console.WriteLine(e);
    throw;
}

```

**Comment by aloneguid** [comment]
Thanks, I've noticed the regression yesterday too. Please hold back one version until I fix it and add some test checks for future.

**Comment by cajuncoding** [comment]
Yep this hit me today too! Thanks for submitting the issue so I could quickly find this 🙏!!!

**Comment by aloneguid** [comment]
@cajuncoding send me the receipt and I'll refund you ;)

**Comment by aloneguid** [comment]
shoudl be fixed in 4.3.0

**Comment by cajuncoding** [comment]
@aloneguid Great! Thanks for the fast response... this resolved the issue in my use cases ✅.
